### ISSUE TITLE
[Bug] MesaNet gradients become inaccurate when the loss is scaled down

### ISSUE THREAD

**Issue by Zhubaoheng** [fla-org/flash-linear-attention#1283]
I ran into this while training a small MesaNet model with a mean-reduced loss. It stayed near random accuracy. Tracing the gradients led to a smaller reproduction using unmodified FLA: **scaling down the gradients passed into `chunk_mesa_net` makes its backward much less accurate, even after dividing out the scale.**

The existing `test_chunk` uses upstream gradients around 1. Keeping its input construction and reducing only the upstream gradient scale is enough to expose the problem. At a scale of `1e-7`, the gradients for `q` and `lamb` have nearly 100% relative error against `naive_mesa_net_exact`.

### Reproduction

The script below keeps the forward inputs fixed, multiplies the gradients of the output and final states by `s`, and divides the returned gradients by `s` before comparing. Mathematically, this should leave the gradients unchanged. Small rounding differences are expected; the large errors below are not.

Here are three columns from its output. Errors are RMSE divided by the reference RMS, as in `fla.utils.assert_close`.

| Upstream scale `s` | `q` gradient error | `lamb` gradient error | Worst error after rescaling before backward* |
|---:|---:|---:|---:|
| 1 | 0.16% | 0.16% | 0.16% |
| 1e-3 | 1.60% | 2.75% | 0.25% |
| 1e-5 | 25.3% | 33.5% | 0.25% |
| 1e-7 | 98.9% | 100% | 0.25% |

*For the last column, I multiply all upstream gradients by a power of two before calling backward, then divide the returned gradients by that factor. This keeps the internal backward computation at roughly unit scale. The script includes this workaround as a comparison.

Only `torch` and `fla` are needed. This takes about one or two minutes on an H100.

<details>
<summary>Reproduction script and full output</summary>

```python
import math

import torch
import torch.nn.functional as F

from fla.ops.mesa_net import chunk_mesa_net, naive_mesa_net_exact

B, T, H, D, dtype, device = 2, 1024, 8, 128, torch.bfloat16, 'cuda'

# inputs: same construction as tests/ops/test_mesa.py::test_chunk (seed 42)
torch.manual_seed(42)
q = torch.rand(B, T, H, D, dtype=dtype) / 10
k = F.normalize(torch.rand(B, T, H, D, dtype=torch.float32), p=2, dim=-1).to(dtype)
v = torch.rand(B, T, H, D, dtype=dtype) / 10
beta = torch.rand(B, T, H, dtype=dtype).sigmoid()
g = torch.rand(B, T, H, dtype=dtype).float().uniform_(0.8, 0.99).log()
lamb = torch.rand(H, D, dtype=dtype).sigmoid() * 0.75 + 0.25
q, k, v, beta, g, lamb = (x.to(device) for x in (q, k, v, beta, g, lamb))
do = torch.rand_like(v)
k0 = F.normalize(torch.rand(B, H, D, device=device, dtype=dtype), dim=-1, p=2)
h_kk_init = (k0.unsqueeze(-1) * k0.unsqueeze(-2)).float()
h_kv_init = torch.rand(B, H, D, D, dtype=torch.float32, device=device)
d_h_kk_final = torch.rand_like(h_kk_init)
d_h_kv_final = torch.rand_like(h_kv_init)

names = ['q', 'k', 'v', 'beta', 'g', 'lamb', 'h_kk_init', 'h_kv_init']
inputs = [q, k, v, beta, g, lamb, h_kk_init, h_kv_init]
tri_x = [x.detach().clone().requires_grad_() for x in inputs]
ref_x = [x.detach().clone().requires_grad_() for x in inputs]
tri_out = chunk_mesa_net(q=tri_x[0], k=tri_x[1], v=tri_x[2], beta=tri_x[3], g=tri_x[4], lamb=tri_x[5],
                         max_CG_iteration=D, h_kk_init=tri_x[6], h_kv_init=tri_x[7], output_final_state=True)
ref_out = naive_mesa_net_exact(q=ref_x[0], k=ref_x[1], v=ref_x[2], beta=ref_x[3], g=ref_x[4], lamb=ref_x[5],
                               h_kk_init=ref_x[6], h_kv_init=ref_x[7])


def err(ref, tri):  # same metric as fla.utils.assert_close
    ref, tri = ref.double(), tri.double()
    return ((ref - tri).pow(2).mean().sqrt() / ref.pow(2).mean().sqrt()).item()


print(f"{'s':>6} | " + ' '.join(f'{n:>9}' for n in names) + ' | prescaled (max)')
for s in [1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7]:
    dO = [(x * s).to(x.dtype) for x in (do, d_h_kk_final, d_h_kv_final)]  # what the kernel actually receives
    ref_g = torch.autograd.grad(ref_out, ref_x, [x.float() / s for x in dO], retain_graph=True)
    tri_g = torch.autograd.grad(tri_out, tri_x, dO, retain_graph=True)
    # Workaround: rescale upstream grads by an exact power of two, then undo it.
    c = 2.0 ** -math.floor(math.log2(max(x.abs().max().item() for x in dO)))
    pre_g = torch.autograd.grad(tri_out, tri_x, [x * c for x in dO], retain_graph=True)
    errs = [err(r, t.double() / s) for r, t in zip(ref_g, tri_g)]
    pre = max(err(r, t.double() / (c * s)) for r, t in zip(ref_g, pre_g))
    print(f'{s:>6.0e} | ' + ' '.join(f'{e:9.2e}' for e in errs) + f' | {pre:9.2e}')
```

Output on H100 with FLA at `954438d`:

```
     s |         q         k         v      beta         g      lamb h_kk_init h_kv_init | prescaled (max)
 1e+00 |  1.63e-03  1.57e-03  1.16e-03  9.14e-04  3.36e-05  1.58e-03  7.19e-04  5.32e-04 |  1.63e-03
 1e-02 |  3.49e-03  2.79e-03  2.36e-03  2.29e-03  6.89e-05  5.48e-03  7.61e-04  5.32e-04 |  2.43e-03
 1e-03 |  1.60e-02  8.96e-03  2.48e-03  2.65e-03  7.91e-04  2.75e-02  2.51e-03  5.32e-04 |  2.49e-03
 1e-04 |  4.56e-02  2.84e-02  2.41e-03  1.74e-02  1.67e-02  5.96e-02  2.35e-02  5.35e-04 |  2.42e-03
 1e-05 |  2.53e-01  1.97e-01  2.51e-03  1.54e-01  1.47e-01  3.35e-01  1.84e-01  1.73e-03 |  2.45e-03
 1e-06 |  6.68e-01  5.90e-01  8.78e-03  4.69e-01  3.45e-01  8.14e-01  6.06e-01  1.68e-02 |  2.50e-03
 1e-07 |  9.89e-01  9.69e-01  1.95e-01  7.41e-01  3.60e-01  1.00e+00  9.84e-01  1.79e-01 |  2.45e-03
```

</details>

I also checked all six shapes from `tests/ops/test_mesa.py::test_chunk`, in both fp16 and bf16, against a float64 version of the exact reference. All 12 configurations show the same pattern. The reference receives the same quantized upstream gradients as FLA, so the comparison does not count information already lost before entering backward as a kernel error.

### What seems to cause it

The backward solves another linear system to propagate gradients through the forward solve. Two implementation details appear to matter:

1. **The CG loop adds a fixed `1e-5` to both denominators.** See [the backward CG loop](https://github.com/fla-org/flash-linear-attention/blob/954438d1fcb5e1bb05c22f9908de9c5c2df74ae5/fla/ops/mesa_net/chunk_cg_solver_bwd.py#L115). When the incoming gradient is small, this constant can dominate the inner products and make the updates far too small. Normalizing each right-hand-side vector before the solve, then restoring its scale afterwards, removes most of the error at `s = 1e-5` in my tests.
2. **The solver output is stored in fp16**, even in the bf16 case: [the call site](https://github.com/fla-org/flash-linear-attention/blob/954438d1fcb5e1bb05c22f9908de9c5c2df74ae5/fla/ops/mesa_net/chunk.py#L143). Small results can lose precision or round to zero. Normalizing the solver input alone does not fix the smallest scales. Some gradients that bypass this solve also degrade, so there appear to be additional precision issues elsewhere in backward.

In the smallest test case at `s = 1e-7`, the solver receives a nonzero fp32 right-hand side with maximum absolute value `1.5e-6`, but returns all zeros. With `gate_range=[1, 1]`, the result is not all zero, but is still inaccurate.

Rescaling before the **whole backward**, rather than just before the solver, keeps the error close to the original test's level across all tested scales. Simply changing the solver output to fp32 is not a drop-in fix: the next kernel fails to compile because it expects `dq` and `k` to have the same dtype.

### Training impact

In my two-layer MQAR model, the loss is averaged over 32,768 queries. I measured gradients at the Mesa layer output around `2–3e-6` initially and `7e-8–3e-7` by step 500, so these small scales occur in an actual training run.

With the original backward, validation accuracy stayed around chance (0.5% at step 1250), and the run failed with non-finite gradients at step 1330. Using backward rescaling and normalized solver inputs, a matched run reached 91.1% at step 1500. This is a single-seed result; the operator reproduction above is the direct evidence for the gradient error. I have not established where the non-finite values first arise.

Would it make sense to add small upstream-gradient scales to `test_chunk` and handle this rescaling inside backward? The wrapper works in these tests without changing the kernels. A solver-level fix may also need changes to the precision of downstream intermediates.

### Environment

- Unmodified FLA at `954438d1fcb5e1bb05c22f9908de9c5c2df74ae5` (includes #1152)
- NVIDIA H100 80GB HBM3
- PyTorch 2.10.0+cu128, Triton 3.6.0
- Python 3.11.15, Linux
